src/core/pull/index.ts

```diff
@@ -0,0 +1,373 @@
+/*!
+ * V4Fire Core
+ * https://github.com/V4Fire/Core
+ *
+ * Released under the MIT license
+ * https://github.com/V4Fire/Core/blob/master/LICENSE
+ */
+
+/**
+ * [[include:core/pull/README.md]]
+ * @packageDocumentation
+ */
+
+import { EventEmitter2 as EventEmitter } from 'eventemitter2';
+import { resolveAfterEvents } from 'core/event';
+
+import SyncPromise from 'core/promise/sync';
+
+import { generate, serialize } from 'core/uuid';
+import { Queue } from 'core/queue';
+
+import type { hook, Options, ReturnType } from 'core/pull/interface';
+import { defaultValue, hashProperty, viewerCount } from 'core/pull/const';
+
+/**
+ * Simple implementation of pull with stack
+ * @typeparam T - pull element
+ */
+export default class Pull<T> {
+	/**
+	 * Factory from constructor argument
+	 */
+	protected objectFactory: (...args: unknown[]) => T;
+
+	/**
+	 * Data structure that contain pull's object
+	 */
+	protected readonly resourceStore: Map<string, T[]> = new Map();
+
+	/**
+	 * Data structure that contain pull's shared object
+	 */
+	protected readonly borrowedResourceStore: Map<string, T> = new Map();
+
+	/**
+	 * Hook that are activated before (free) function
+	 *
+	 * @param value - value from free(value)
+	 * @param pull - this pull
+	 * @param args - params that are given in free(value,...args)
+	 */
+	protected onFree?: hook<T>;
+
+	/**
+	 * Hook that are activated before this.take or this.takeOrCreate
```

Comment:
```suggestion
	 * Hook that are activated before `this.take` or `this.takeOrCreate`
```

Везде врапь в \`\` такие штуки